### PR TITLE
Refactor config loading and add loading screen

### PR DESCRIPTION
**Pull request by zbettenbuk**

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -116,7 +116,7 @@ const config = {
 module.exports = [
     Object.assign({}, config, {
         entry: {
-            'app.bundle': './app.js',
+            'app.bundle': [ '@babel/polyfill', './app.js' ],
```
Comment (virtuacoplenny) [inline webpack.config.js]: polyfill is for async/await? We're going to start using that? (If so I'll add it to spot as well.) 

Also, should we try to target the async/away polyfills and not bring in the package? Admittedly, I haven't looked into it myself. Another concern I have would be how to get the polyfills to the other web bundles below. I'm also tempted to recommend creating yet another bundle so the bundles below can import the polyfills through a script tag.

File: react/features/app/actions.js [hunk 2]
```diff
@@ -31,176 +31,72 @@ declare var APP: Object;
...
+        let protocol = location.protocol.toLowerCase();
+        const { contextRoot, host, room } = location;
 
-    /* eslint-disable no-param-reassign */
+        // The React Native app supports an app-specific scheme which is sure to not
+        // be supported by fetch (or whatever loadConfig utilizes).
+        protocol !== 'http:' && protocol !== 'https:' && (protocol = 'https:');
 
-    protocol = protocol.toLowerCase();
+        // TDOO userinfo
```
Comment (virtuacoplenny) [inline react/features/app/actions.js]: I realize this isn't your code but could you expand up on this comment a bit if you understand what it's for? I guess at least fix the typo so it says TODO.

File: react/features/app/actions.js [hunk 3]
```diff
@@ -31,176 +31,72 @@ declare var APP: Object;
...
- * state.
- * @param {Object} newLocation - The location URI to navigate to. The value
- * cannot be undefined and is assumed to have all properties such as
- * {@code host}, {@code contextRoot}, and {@code room} defined. Depending on the
- * property, it may have a value equal to {@code undefined} and that may be
- * acceptable.
- * @private
- * @returns {Promise<void>}
- */
-function _appNavigateToMandatoryLocation(
```
Comment (virtuacoplenny) [inline react/features/app/actions.js]: Could you provide an explanation for why you de-encapsulated parts of the config load logic?

In the past I had trouble understanding this logic and specifically the optional/mandatory terminology; I didn't have context and like an amateur didn't ask for it. I think because of that missing context this might be the rare case where I prefer having one bigger function like you did, so that my mental context stays within the same function and its scope.

File: react/features/app/actions.js [hunk 4]
```diff
@@ -31,176 +31,72 @@ declare var APP: Object;
...
-        dispatch: Dispatch<any>,
-        getState: Function,
-        { contextRoot, host, protocol, room }) {
-    // XXX As the mobile/React Native app does not employ config on the
-    // WelcomePage, do not download config.js from the deployment when
-    // navigating to the WelcomePage - the perceived/visible navigation will be
-    // faster.
-    if (!room && typeof APP === 'undefined') {
-        return Promise.resolve();
-    }
```
Comment (virtuacoplenny) [inline react/features/app/actions.js]: This no-oping for mobile has been removed? So now mobile loads config, yeah?

File: react/features/app/actions.js [hunk 5]
```diff
@@ -31,176 +31,72 @@ declare var APP: Object;
...
+                dispatch(setConfig(config));
 
-            if (config) {
-                return config;
+                dispatch(setRoom(room));
             }
 
-            throw error;
-        });
+            dispatch(loadConfigError(new Error('Config no longer needed!'), locationURL));
```
Comment (virtuacoplenny) [inline react/features/app/actions.js]: Should the dispatching of loadConfigError be guarded behind some kind of check? Seems like this might always fire if config is defined.

File: webpack.config.js (hunk 1 above)
Comment (zbettenbuk) [reply webpack.config.js]: Yes, it was needed for async/await. I hope we can get rid of some additional manual polyfills with it as well, but it's not the scope of this PR

File: react/features/app/actions.js (hunk 5 above)
Comment (virtuacoplenny) [reply react/features/app/actions.js]: Or maybe it should be removed because you already dispatch the error above?

File: react/features/app/actions.js (hunk 2 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: Lyubo wrote it 2 years ago... I think it's not really meaningful anymore :)

File: react/features/app/actions.js (hunk 3 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: Yes, the whole point was (wishful) readability. Removing things like functions with a single line invoking another function, and the implicitly introduced optional/mandatory navigation (which is a bad naming anyway).

File: react/features/app/actions.js (hunk 4 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: Yes, we have plans with it for later

File: react/features/app/actions.js (hunk 5 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: No, it's a bug... I used to have return statements there, so it wasn't invoked after the ```if``` but now they are gone. Fixed it, thanks!

File: react/features/app/actions.js [hunk 6]
```diff
@@ -31,176 +31,70 @@ declare var APP: Object;
...
-            const config = restoreConfig(baseURL);
+            if (getState()['features/base/config'].locationURL === locationURL) {
+                dispatch(setLocationURL(locationURL));
+                dispatch(setConfig(config));
 
-            if (config) {
-                return config;
+                dispatch(setRoom(room));
+            } else {
+                dispatch(loadConfigError(new Error('Config no longer needed!'), locationURL));
```
Comment (virtuacoplenny) [inline react/features/app/actions.js]: Does dispatching loadConfigError here again conflict with the dispatch of loadConfigError above?

Comment (paweldomas) [reply react/features/app/actions.js]: These are two different kind of failures. This one is when multiple "load config" tasks are started and it gets outdated - only the most recent one is to be accepted in the end. The other one above if it just fails to fetch.

Comment (virtuacoplenny) [reply react/features/app/actions.js]: So there isn't a worry about the error here overriding the stored error above?

Comment (paweldomas) [reply react/features/app/actions.js]: Nope, the error will be stored only for the current request:
https://github.com/jitsi/jitsi-meet/blob/master/react/features/base/config/reducer.js#L70

**Comment by virtuacoplenny** [comment]
Should https://github.com/jitsi/jitsi-meet/pull/3981 be closed?

File: react/features/app/actions.js [hunk 7]
```diff
@@ -31,176 +31,70 @@ declare var APP: Object;
...
-            // setters in order to reduce the risks of inconsistent state.
-            location.hostname = defaultLocation.hostname;
-            location.pathname
-                = defaultLocation.pathname + location.pathname.substr(1);
-            location.port = defaultLocation.port;
-            location.protocol = defaultLocation.protocol;
-        } else {
-            // eslint-disable-next-line no-param-reassign
-            location = defaultLocation;
+    return async(dispatch: Dispatch<any>, getState: Function) => {
```
Comment (saghul) [inline react/features/app/actions.js]: style: space after async

File: react/features/app/actions.js [hunk 8]
```diff
@@ -31,176 +31,70 @@ declare var APP: Object;
...
+
+                // FIXME Turn location's host, hostname, and port properties into
+                // setters in order to reduce the risks of inconsistent state.
+                location.hostname = defaultLocation.hostname;
+                location.pathname
+                    = defaultLocation.pathname + location.pathname.substr(1);
+                location.port = defaultLocation.port;
+                location.protocol = defaultLocation.protocol;
+            } else {
+                // eslint-disable-next-line no-param-reassign
```
Comment (saghul) [inline react/features/app/actions.js]: is this needed? `location is no longer a parameter

File: react/features/app/actions.js [hunk 9]
```diff
@@ -31,176 +31,70 @@ declare var APP: Object;
...
 
-    protocol = protocol.toLowerCase();
+        let protocol = location.protocol.toLowerCase();
+        const { contextRoot, host, room } = location;
 
-    // The React Native app supports an app-specific scheme which is sure to not
-    // be supported by fetch (or whatever loadConfig utilizes).
-    protocol !== 'http:' && protocol !== 'https:' && (protocol = 'https:');
+        // The React Native app supports an app-specific scheme which is sure to not
+        // be supported by fetch (or whatever loadConfig utilizes).
```
Comment (saghul) [inline react/features/app/actions.js]: can you remove the `or whatever loadConfig utilizes` part? It doesn't add anything.

File: react/features/app/actions.js [hunk 10]
```diff
@@ -31,176 +31,70 @@ declare var APP: Object;
...
-    // XXX In order to support multiple shards, tell the room to the deployment.
-    room && (url += `?room=${room.toLowerCase()}`);
+        let config;
 
-    /* eslint-enable no-param-reassign */
+        try {
+            config = await loadConfig(url);
+        } catch (error) {
+            config = restoreConfig(baseURL);
+            dispatch(loadConfigError(error, locationURL));
```
Comment (saghul) [inline react/features/app/actions.js]: I think we only want to this if we got no config after the restore, not unconditionally

File: webpack.config.js (hunk 1 above)
Comment (saghul) [reply webpack.config.js]: I'd say it's well in scope of this PR. We have removed babel polyfill previously and I don't want it back. Let's look at the options. There seem to be a number of them. I'm going to try some.

File: react/features/authentication/components/LoginDialog.native.js [hunk 11]
```diff
@@ -281,7 +281,7 @@ class LoginDialog extends Component<Props, State> {
         if (conference) {
             r = dispatch(authenticateAndUpgradeRole(jid, password, conference));
         } else {
-            r = dispatch(connect(jid, password));
```
Comment (saghul) [inline react/features/authentication/components/LoginDialog.native.js]: Why this change? At a quick glance this might break authentication on some scenarios, no?

File: react/features/invite/functions.js [hunk 12]
```diff
@@ -266,7 +266,7 @@ export function getInviteTypeCounts(inviteItems: Array<Object> = []) {
  */
 export function invitePeopleAndChatRooms( // eslint-disable-line max-params
         inviteServiceUrl: string,
-        inviteUrl: string,
+        inviteUrl: ?string,
```
Comment (saghul) [inline react/features/invite/functions.js]: how can this be optional? where would we be inviting users? I think the fix must be on the caller side, to make sure we always pass a string

File: react/features/mobile/call-integration/middleware.js [hunk 13]
```diff
@@ -228,14 +228,16 @@ function _conferenceWillJoin({ dispatch, getState }, next, action) {
     const state = getState();
     const { callHandle, callUUID } = state['features/base/config'];
     const url = getInviteURL(state);
-    const handle = callHandle || url.toString();
```
Comment (saghul) [inline react/features/mobile/call-integration/middleware.js]: same as above, this change is incorrect: a conference will always have a URL

File: react/features/app/actions.js (hunk 7 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: Nope, eslint doesn't let me

File: react/features/authentication/components/LoginDialog.native.js (hunk 11 above)
Comment (zbettenbuk) [reply react/features/authentication/components/LoginDialog.native.js]: No, the connect function doesn't accept any parameter

File: react/features/invite/functions.js (hunk 12 above)
Comment (zbettenbuk) [reply react/features/invite/functions.js]: Probably, but then we need to change the inviteUrl generation logic, because that can explicitly return undefined (I guess there was a reason for it)

File: react/features/mobile/call-integration/middleware.js (hunk 13 above)
Comment (zbettenbuk) [reply react/features/mobile/call-integration/middleware.js]: This is not my change

File: react/features/app/actions.js (hunk 7 above)
Comment (saghul) [reply react/features/app/actions.js]: The we need a new rule or something. `async` needs a space, otherwise `async function foo() { ... }` cannot be written.

File: react/features/authentication/components/LoginDialog.native.js (hunk 11 above)
Comment (saghul) [reply react/features/authentication/components/LoginDialog.native.js]: https://github.com/jitsi/jitsi-meet/blob/7a677ead93d191f5dac5df909f3d948efbeec76e/react/features/base/connection/actions.native.js#L78 Isn't it this one?

File: react/features/invite/functions.js (hunk 12 above)
Comment (saghul) [reply react/features/invite/functions.js]: Then let's change that, as it is this change just hides the problem.

File: react/features/mobile/call-integration/middleware.js (hunk 13 above)
Comment (saghul) [reply react/features/mobile/call-integration/middleware.js]: How so? You care modyfying this thing now.

File: webpack.config.js (hunk 1 above)
Comment (hristoterezov) [reply webpack.config.js]: I'm wondering if it is better to introduce async/await trough separate PR. This way you can refactor in new async/await PR the other places of the code that you mentioned. In this PR you are using it only at one place and we will have to add the polyfill dependancy. I'm not against starting using async/await but in this case it feels a little bit that you actually don't really need it in the PR but you have added it only to be able to use it later....

Comment (zbettenbuk) [reply webpack.config.js]: @saghul I'm waiting for your suggestions around this. Otherwise I'd go with this implementation.

@hristoterezov Yeah, you're probably right, although I did all this "cleaning up" to have better readability, that involves async/await, If you feel a strong urge to have it in a separate commit, I'm happy to do it, but I'd be happier to leave it like this, because then I had to implement another logic here with promises :)

File: react/features/app/actions.js (hunk 7 above)
Comment (zbettenbuk) [reply react/features/app/actions.js]: No, that would work, this is a separate eslint rule, see PR: https://github.com/jitsi/eslint-config-jitsi/pull/10

File: react/features/authentication/components/LoginDialog.native.js (hunk 11 above)
Comment (zbettenbuk) [reply react/features/authentication/components/LoginDialog.native.js]: According to flow, not, but you're right. Thanks

File: webpack.config.js (hunk 1 above)
Comment (saghul) [reply webpack.config.js]: @zbettenbuk I'll look into it, but you should too.

@hristoterezov Adding async/await was the heart of this PR, adding support for it without a single usage would be weird IMHO.

File: react/features/app/actions.js (hunk 2 above)
Comment (saghul) [reply react/features/app/actions.js]: https://en.wikipedia.org/wiki/Uniform_Resource_Identifier#Definition

File: package.json [hunk 14]
```diff
@@ -99,7 +100,7 @@
     "clean-css": "3.4.25",
     "css-loader": "0.28.7",
     "eslint": "5.6.1",
-    "eslint-config-jitsi": "github:jitsi/eslint-config-jitsi#1.0.0",
+    "eslint-config-jitsi": "github:jitsi/eslint-config-jitsi#5f9fe4fd00be471ab6f5d06f77ffc90bba0853e8",
```
Comment (saghul) [inline package.json]: Make a new release please.

Comment (zbettenbuk) [reply package.json]: Does it really make sense to make a new release every time we change a single line?

Comment (saghul) [reply package.json]: Yes.

File: webpack.config.js (hunk 1 above)
Comment (saghul) [reply webpack.config.js]: Ok, I found a solution! The trick is to let Babel know what browsers we support: https://github.com/saghul/jitsi-meet/blob/8f056c278bc5e6c23fd146cb1296354f5e13c9f8/webpack.config.js#L61-L68

Then we can use async natively.

Here is a commit which removes the polyfill and uses that approach: https://github.com/saghul/jitsi-meet/commit/8f056c278bc5e6c23fd146cb1296354f5e13c9f8

File: react/features/app/actions.js (hunk 10 above)
Comment (saghul) [reply react/features/app/actions.js]: This is not fixed yet, but I see it marked as resolved.  Fell trough the cracks?

File: package.json (hunk 14 above)
Comment (saghul) [reply package.json]: can you please use the tag reference here?

File: react/features/overlay/components/AbstractPageReloadOverlay.js [hunk 15]
```diff
@@ -14,7 +14,7 @@ import {
     isFatalJitsiConnectionError
 } from '../../base/lib-jitsi-meet';
 
-import ReloadButton from './ReloadButton';
```
Comment (saghul) [inline react/features/overlay/components/AbstractPageReloadOverlay.js]: you sure this doesn't run on mobile?

File: react/features/overlay/components/native/LoadConfigOverlay.js [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+ * Maps part of the Redux state to the props of this component.
+ *
+ * @param {Object} state - The Redux state.
+ * @returns {{
+ *     _styles: StyleType
+ * }}
+ */
+function _mapStateToProps(state) {
+    return {
+        _styles: ColorSchemeRegistry.get(state, 'Overlay')
```
Comment (saghul) [inline react/features/overlay/components/native/LoadConfigOverlay.js]: Lets call this LoadConfigOverlay, in case we need / want different colors for different overlays.

File: webpack.config.js [hunk 17]
```diff
@@ -56,7 +56,19 @@ const config = {
...
                         // so that webpack may do tree shaking.
-                        { modules: false }
+                        {
+                            modules: false,
+
+                            // Specify our target browsers so no transpiling is
+                            // done unnecessarily. For browsers not specified
+                            // here, the ES2015+ profile will be used.
+                            targets: {
+                                chrome: 58,
```
Comment (saghul) [inline webpack.config.js]: can you please add `electron: 2` here? I forgot!

File: react/features/overlay/components/AbstractPageReloadOverlay.js (hunk 15 above)
Comment (zbettenbuk) [reply react/features/overlay/components/AbstractPageReloadOverlay.js]: No, and it has a web render method (with divs and all)

Comment (zbettenbuk) [reply react/features/overlay/components/AbstractPageReloadOverlay.js]: Btw it's not good that we have it referenced here in the abstract class, but I just really didn't want to push this PR as far as fixing even this
